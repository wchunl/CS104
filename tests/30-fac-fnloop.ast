ROOT "" (0.0.0)
|   TOK_FUNCTION "" (6.22.1)
|   |   TOK_TYPE_ID "" (6.22.1)
|   |   |   TOK_VOID "void" (6.22.1)
|   |   |   TOK_IDENT "fail" (6.22.6)
|   |   TOK_PARAM "(" (6.22.11)
|   |   |   TOK_TYPE_ID "" (6.22.12)
|   |   |   |   TOK_STRING "string" (6.22.12)
|   |   |   |   TOK_IDENT "expr" (6.22.19)
|   |   |   TOK_TYPE_ID "" (6.22.25)
|   |   |   |   TOK_STRING "string" (6.22.25)
|   |   |   |   TOK_IDENT "file" (6.22.32)
|   |   |   TOK_TYPE_ID "" (6.22.38)
|   |   |   |   TOK_INT "int" (6.22.38)
|   |   |   |   TOK_IDENT "line" (6.22.42)
|   TOK_FUNCTION "" (6.24.1)
|   |   TOK_TYPE_ID "" (6.24.1)
|   |   |   TOK_VOID "void" (6.24.1)
|   |   |   TOK_IDENT "putchr" (6.24.6)
|   |   TOK_PARAM "(" (6.24.13)
|   |   |   TOK_TYPE_ID "" (6.24.14)
|   |   |   |   TOK_INT "int" (6.24.14)
|   |   |   |   TOK_IDENT "chr" (6.24.18)
|   TOK_FUNCTION "" (6.25.1)
|   |   TOK_TYPE_ID "" (6.25.1)
|   |   |   TOK_VOID "void" (6.25.1)
|   |   |   TOK_IDENT "putint" (6.25.6)
|   |   TOK_PARAM "(" (6.25.13)
|   |   |   TOK_TYPE_ID "" (6.25.14)
|   |   |   |   TOK_INT "int" (6.25.14)
|   |   |   |   TOK_IDENT "num" (6.25.18)
|   TOK_FUNCTION "" (6.26.1)
|   |   TOK_TYPE_ID "" (6.26.1)
|   |   |   TOK_VOID "void" (6.26.1)
|   |   |   TOK_IDENT "putstr" (6.26.6)
|   |   TOK_PARAM "(" (6.26.13)
|   |   |   TOK_TYPE_ID "" (6.26.14)
|   |   |   |   TOK_STRING "string" (6.26.14)
|   |   |   |   TOK_IDENT "str" (6.26.21)
|   TOK_FUNCTION "" (6.28.1)
|   |   TOK_TYPE_ID "" (6.28.1)
|   |   |   TOK_INT "int" (6.28.1)
|   |   |   TOK_IDENT "getchr" (6.28.5)
|   |   TOK_PARAM "(" (6.28.11)
|   TOK_FUNCTION "" (6.29.1)
|   |   TOK_TYPE_ID "" (6.29.1)
|   |   |   TOK_STRING "string" (6.29.1)
|   |   |   TOK_IDENT "getstr" (6.29.8)
|   |   TOK_PARAM "(" (6.29.14)
|   TOK_FUNCTION "" (6.30.1)
|   |   TOK_TYPE_ID "" (6.30.1)
|   |   |   TOK_STRING "string" (6.30.1)
|   |   |   TOK_IDENT "getln" (6.30.8)
|   |   TOK_PARAM "(" (6.30.13)
|   TOK_FUNCTION "" (7.8.1)
|   |   TOK_TYPE_ID "" (7.8.1)
|   |   |   TOK_INT "int" (7.8.1)
|   |   |   TOK_IDENT "fac" (7.8.5)
|   |   TOK_PARAM "(" (7.8.9)
|   |   |   TOK_TYPE_ID "" (7.8.10)
|   |   |   |   TOK_INT "int" (7.8.10)
|   |   |   |   TOK_IDENT "n" (7.8.14)
|   |   TOK_BLOCK "{" (7.8.17)
|   |   |   TOK_VARDECL "=" (7.9.10)
|   |   |   |   TOK_INT "int" (7.9.4)
|   |   |   |   |   TOK_TYPE_ID "f" (7.9.8)
|   |   |   |   TOK_INTCON "1" (7.9.12)
|   |   |   |   TOK_WHILE "while" (7.10.4)
|   |   |   |   |   '>' ">" (7.10.13)
|   |   |   |   |   |   TOK_IDENT "n" (7.10.11)
|   |   |   |   |   |   TOK_INTCON "1" (7.10.15)
|   |   |   |   |   TOK_BLOCK "{" (7.10.18)
|   |   |   |   |   |   '=' "=" (7.11.9)
|   |   |   |   |   |   |   TOK_IDENT "f" (7.11.7)
|   |   |   |   |   |   |   '*' "*" (7.11.13)
|   |   |   |   |   |   |   |   TOK_IDENT "f" (7.11.11)
|   |   |   |   |   |   |   |   TOK_IDENT "n" (7.11.15)
|   |   |   |   |   |   |   '=' "=" (7.12.9)
|   |   |   |   |   |   |   |   TOK_IDENT "n" (7.12.7)
|   |   |   |   |   |   |   |   '-' "-" (7.12.13)
|   |   |   |   |   |   |   |   |   TOK_IDENT "n" (7.12.11)
|   |   |   |   |   |   |   |   |   TOK_INTCON "1" (7.12.15)
|   |   |   |   TOK_RETURN "return" (7.14.4)
|   |   |   |   |   TOK_IDENT "f" (7.14.11)
|   TOK_FUNCTION "" (7.17.1)
|   |   TOK_TYPE_ID "" (7.17.1)
|   |   |   TOK_INT "int" (7.17.1)
|   |   |   TOK_IDENT "main" (7.17.5)
|   |   TOK_PARAM "(" (7.17.9)
|   |   TOK_BLOCK "{" (7.17.12)
|   |   |   TOK_VARDECL "=" (7.18.10)
|   |   |   |   TOK_INT "int" (7.18.4)
|   |   |   |   |   TOK_TYPE_ID "n" (7.18.8)
|   |   |   |   TOK_INTCON "1" (7.18.12)
|   |   |   |   TOK_WHILE "while" (7.19.4)
|   |   |   |   |   TOK_LE "<=" (7.19.13)
|   |   |   |   |   |   TOK_IDENT "n" (7.19.11)
|   |   |   |   |   |   TOK_INTCON "5" (7.19.16)
|   |   |   |   |   TOK_BLOCK "{" (7.19.19)
|   |   |   |   |   |   TOK_CALL "(" (7.20.14)
|   |   |   |   |   |   |   TOK_IDENT "putint" (7.20.7)
|   |   |   |   |   |   |   TOK_CALL "(" (7.20.19)
|   |   |   |   |   |   |   |   TOK_IDENT "fac" (7.20.15)
|   |   |   |   |   |   |   |   TOK_IDENT "n" (7.20.20)
|   |   |   |   |   |   |   TOK_CALL "(" (7.21.14)
|   |   |   |   |   |   |   |   TOK_IDENT "putchr" (7.21.7)
|   |   |   |   |   |   |   |   TOK_CHARCON "'\n'" (7.21.15)
|   |   |   |   |   |   |   '=' "=" (7.22.9)
|   |   |   |   |   |   |   |   TOK_IDENT "n" (7.22.7)
|   |   |   |   |   |   |   |   '+' "+" (7.22.13)
|   |   |   |   |   |   |   |   |   TOK_IDENT "n" (7.22.11)
|   |   |   |   |   |   |   |   |   TOK_INTCON "1" (7.22.15)
|   |   |   |   TOK_RETURN "return" (7.24.4)
|   |   |   |   |   TOK_INTCON "0" (7.24.11)
