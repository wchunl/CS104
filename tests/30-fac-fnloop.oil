fail:     .function
          .param ptr expr
          .param ptr file
          .param int line
          .end
putchr:   .function
          .param int chr
          .end
putint:   .function
          .param int num
          .end
putstr:   .function
          .param ptr str
          .end
getchr:   .function int
          .end
getstr:   .function
          .end
getln:    .function
          .end
fac:      .function int
          .param int n
          .local int f
          f = 1
.wh0:     $t0:i = n > 1
          goto .od0 if not $t0:i
.do0:     $t1:i = f * n
          f = $t1:i
          $t2:i = n - 1
          n = $t2:i
.od0:     return f
          return 
          .end
main:     .function int
          .local int n
          n = 1
.wh1:     $t3:i = n <= 5
          goto .od1 if not $t3:i
.do1:     fac(n)
          putint($t4:i)
          putchr('\n')
          $t5:i = n + 1
          n = $t5:i
.od1:     return 0
          return 
          .end
