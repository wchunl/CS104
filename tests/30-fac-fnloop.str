string_set[   0]:    4706206694200543916 0x1419c28->","
                   16579916199511856337 0x141b0b8->"getln"
string_set[   1]:    2297668033614959926 0x141c7a8->"0"
                    6637313742931709005 0x141bd28->"*"
string_set[   3]:    7995238595416485409 0x141c9c8->"5"
                   11868120863286193964 0x141c408->"main"
                    9611428142580219387 0x14155a8->"fail"
string_set[   4]:      21606623943142546 0x141b778->"{"
                    4017852018407011111 0x1419d18->"file"
string_set[   6]:   13686721134187162920 0x141bfc8->"}"
string_set[   8]:    1370730858159036685 0x141ca68->"+"
                    9386170080689500517 0x1419af8->"expr"
string_set[   9]:   14526079891305879776 0x141bc38->">"
                   10319632053171904911 0x141baf8->"while"
string_set[  12]:   15816506932648033945 0x1415228->"("
string_set[  17]:   15430489877420048373 0x141c928->"<="
                   18413295556186398326 0x141a118->"putint"
string_set[  18]:    7211130878751977441 0x141a248->"putchr"
                   12679825168387489523 0x141a728->"num"
string_set[  21]:    8968846175329310707 0x141b868->"f"
string_set[  23]:   11166701219533395393 0x141ae78->"getstr"
                   12868445110721718657 0x1419d38->")"
string_set[  28]:   14635384904416252381 0x141ba68->"return"
string_set[  29]:   12291392672316406109 0x1419ec8->"line"
string_set[  31]:    6192799629375248351 0x141ce08->"'\n'"
                    6142509188972423790 0x1415588->""
                   13348923631108115385 0x141a508->"putstr"
string_set[  32]:   11597697714117577063 0x141b928->"="
string_set[  33]:    1583576014053361350 0x141a8a8->"str"
string_set[  34]:    7372855250431704371 0x141c168->"-"
                   13693284975010602715 0x141b2d8->"fac"
string_set[  38]:   12597418120160561823 0x1419b68->"int"
string_set[  39]:    7296690267450279874 0x141a388->"chr"
string_set[  40]:   11911796782241914524 0x1419688->"void"
string_set[  42]:    5344662657899890615 0x1419ee8->";"
string_set[  44]:   13497578857207247412 0x1419a28->"string"
string_set[  45]:   10159970873491820195 0x141b9c8->"1"
                    9826907452181669771 0x141b538->"n"
string_set[  46]:    9201800968376702500 0x141ac58->"getchr"
load_factor = 0.830
bucket_count = 47
max_bucket_size = 3
