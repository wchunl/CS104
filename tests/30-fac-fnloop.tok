#  1 "30-fac-fnloop.oc"
#  1 "<built-in>"
#  1 "<command-line>"
#  1 "30-fac-fnloop.oc"
#  1 "oclib.h"
# 22 "oclib.h"
   6 22.00001 261 TOK_VOID      void
   6 22.00006 282 TOK_IDENT     fail
   6 22.00011  40 '('           (
   6 22.00012 263 TOK_STRING    string
   6 22.00019 282 TOK_IDENT     expr
   6 22.00023  44 ','           ,
   6 22.00025 263 TOK_STRING    string
   6 22.00032 282 TOK_IDENT     file
   6 22.00036  44 ','           ,
   6 22.00038 262 TOK_INT       int
   6 22.00042 282 TOK_IDENT     line
   6 22.00046  41 ')'           )
   6 22.00047  59 ';'           ;
   6 24.00001 261 TOK_VOID      void
   6 24.00006 282 TOK_IDENT     putchr
   6 24.00013  40 '('           (
   6 24.00014 262 TOK_INT       int
   6 24.00018 282 TOK_IDENT     chr
   6 24.00021  41 ')'           )
   6 24.00022  59 ';'           ;
   6 25.00001 261 TOK_VOID      void
   6 25.00006 282 TOK_IDENT     putint
   6 25.00013  40 '('           (
   6 25.00014 262 TOK_INT       int
   6 25.00018 282 TOK_IDENT     num
   6 25.00021  41 ')'           )
   6 25.00022  59 ';'           ;
   6 26.00001 261 TOK_VOID      void
   6 26.00006 282 TOK_IDENT     putstr
   6 26.00013  40 '('           (
   6 26.00014 263 TOK_STRING    string
   6 26.00021 282 TOK_IDENT     str
   6 26.00024  41 ')'           )
   6 26.00025  59 ';'           ;
   6 28.00001 262 TOK_INT       int
   6 28.00005 282 TOK_IDENT     getchr
   6 28.00011  40 '('           (
   6 28.00012  41 ')'           )
   6 28.00013  59 ';'           ;
   6 29.00001 263 TOK_STRING    string
   6 29.00008 282 TOK_IDENT     getstr
   6 29.00014  40 '('           (
   6 29.00015  41 ')'           )
   6 29.00016  59 ';'           ;
   6 30.00001 263 TOK_STRING    string
   6 30.00008 282 TOK_IDENT     getln
   6 30.00013  40 '('           (
   6 30.00014  41 ')'           )
   6 30.00015  59 ';'           ;
#  7 "30-fac-fnloop.oc"
   7 8.00001 262 TOK_INT       int
   7 8.00005 282 TOK_IDENT     fac
   7 8.00009  40 '('           (
   7 8.00010 262 TOK_INT       int
   7 8.00014 282 TOK_IDENT     n
   7 8.00015  41 ')'           )
   7 8.00017 123 '{'           {
   7 9.00004 262 TOK_INT       int
   7 9.00008 282 TOK_IDENT     f
   7 9.00010  61 '='           =
   7 9.00012 283 TOK_INTCON    1
   7 9.00013  59 ';'           ;
   7 10.00004 267 TOK_WHILE     while
   7 10.00010  40 '('           (
   7 10.00011 282 TOK_IDENT     n
   7 10.00013  62 '>'           >
   7 10.00015 283 TOK_INTCON    1
   7 10.00016  41 ')'           )
   7 10.00018 123 '{'           {
   7 11.00007 282 TOK_IDENT     f
   7 11.00009  61 '='           =
   7 11.00011 282 TOK_IDENT     f
   7 11.00013  42 '*'           *
   7 11.00015 282 TOK_IDENT     n
   7 11.00016  59 ';'           ;
   7 12.00007 282 TOK_IDENT     n
   7 12.00009  61 '='           =
   7 12.00011 282 TOK_IDENT     n
   7 12.00013  45 '-'           -
   7 12.00015 283 TOK_INTCON    1
   7 12.00016  59 ';'           ;
   7 13.00004 125 '}'           }
   7 14.00004 268 TOK_RETURN    return
   7 14.00011 282 TOK_IDENT     f
   7 14.00012  59 ';'           ;
   7 15.00001 125 '}'           }
   7 17.00001 262 TOK_INT       int
   7 17.00005 282 TOK_IDENT     main
   7 17.00009  40 '('           (
   7 17.00010  41 ')'           )
   7 17.00012 123 '{'           {
   7 18.00004 262 TOK_INT       int
   7 18.00008 282 TOK_IDENT     n
   7 18.00010  61 '='           =
   7 18.00012 283 TOK_INTCON    1
   7 18.00013  59 ';'           ;
   7 19.00004 267 TOK_WHILE     while
   7 19.00010  40 '('           (
   7 19.00011 282 TOK_IDENT     n
   7 19.00013 278 TOK_LE        <=
   7 19.00016 283 TOK_INTCON    5
   7 19.00017  41 ')'           )
   7 19.00019 123 '{'           {
   7 20.00007 282 TOK_IDENT     putint
   7 20.00014  40 '('           (
   7 20.00015 282 TOK_IDENT     fac
   7 20.00019  40 '('           (
   7 20.00020 282 TOK_IDENT     n
   7 20.00021  41 ')'           )
   7 20.00022  41 ')'           )
   7 20.00023  59 ';'           ;
   7 21.00007 282 TOK_IDENT     putchr
   7 21.00014  40 '('           (
   7 21.00015 284 TOK_CHARCON   '\n'
   7 21.00019  41 ')'           )
   7 21.00020  59 ';'           ;
   7 22.00007 282 TOK_IDENT     n
   7 22.00009  61 '='           =
   7 22.00011 282 TOK_IDENT     n
   7 22.00013  43 '+'           +
   7 22.00015 283 TOK_INTCON    1
   7 22.00016  59 ';'           ;
   7 23.00004 125 '}'           }
   7 24.00004 268 TOK_RETURN    return
   7 24.00011 283 TOK_INTCON    0
   7 24.00012  59 ';'           ;
   7 25.00001 125 '}'           }
