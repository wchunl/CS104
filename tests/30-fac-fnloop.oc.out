; 0.  1: # 1 "30-fac-fnloop.oc"; 1.  0: 
; 1.  1: # 1 "<built-in>"
; 2.  1: # 1 "<command-line>"
; 3.  1: # 1 "30-fac-fnloop.oc"
; 4.  1: 
; 4.  2: 
; 4.  3: 
; 4.  4: 
; 4.  5: 
; 4.  6: # 1 "oclib.h" 1
; 5.  1: # 22 "oclib.h"
; 6. 22: void fail (string expr, string file, int line);
; 6. 23: 
; 6. 24: void putchr (int chr);
; 6. 25: void putint (int num);
; 6. 26: void putstr (string str);
; 6. 27: 
; 6. 28: int getchr();
; 6. 29: string getstr();
; 6. 30: string getln();
; 6. 31: # 7 "30-fac-fnloop.oc" 2
; 7.  7: 
; 7.  8: int fac (int n) {
; 7.  9:    int f = 1;
; 7. 10:    while (n > 1) {
; 7. 11:       f = f * n;
; 7. 12:       n = n - 1;
; 7. 13:    }
; 7. 14:    return f;
; 7. 15: }
; 7. 16: 
; 7. 17: int main() {
; 7. 18:    int n = 1;
; 7. 19:    while (n <= 5) {
; 7. 20:       putint (fac (n));
; 7. 21:       putchr ('\n');
; 7. 22:       n = n + 1;
; 7. 23:    }
; 7. 24:    return 0;
; 7. 25: }
hello